builtin/providers/azurerm/resource_arm_app_gateway.go

```diff
@@ -0,0 +1,1667 @@
+
+		setting := network.ApplicationGatewayBackendHTTPSettings{
+			Name: &name,
+			ApplicationGatewayBackendHTTPSettingsPropertiesFormat: &network.ApplicationGatewayBackendHTTPSettingsPropertiesFormat{
+				Port:                &port,
+				Protocol:            network.ApplicationGatewayProtocol(protocol),
+				CookieBasedAffinity: network.ApplicationGatewayCookieBasedAffinity(cookieBasedAffinity),
+				RequestTimeout:      &requestTimeout,
+			},
+		}
+
+		if data["authentication_certificate"] != nil {
+			authCerts := data["authentication_certificate"].([]interface{})
+			authCertSubResources := make([]network.SubResource, 0, len(authCerts))
+
+			for _, rawAuthCert := range authCerts {
+				authCert := rawAuthCert.(map[string]interface{})
+				authCertID := fmt.Sprintf("%s/authenticationCertificates/%s", gatewayID, authCert["name"])
+				authCertSubResource := network.SubResource{
+					ID: &authCertID,
+				}
+
+				authCertSubResources = append(authCertSubResources, authCertSubResource)
+			}
+
+			setting.ApplicationGatewayBackendHTTPSettingsPropertiesFormat.AuthenticationCertificates = &authCertSubResources
+		}
+
+		probeName := data["probe_name"].(string)
+		if probeName != "" {
+			probeID := fmt.Sprintf("%s/probes/%s", gatewayID, probeName)
+			setting.ApplicationGatewayBackendHTTPSettingsPropertiesFormat.Probe = &network.SubResource{
+				ID: &probeID,
+			}
+		}
+
+		backendSettings = append(backendSettings, setting)
+	}
+
+	return &backendSettings
+}
+
+func expandAppGatewayHTTPListeners(d *schema.ResourceData, gatewayID string) *[]network.ApplicationGatewayHTTPListener {
+	configs := d.Get("http_listener").([]interface{})
+	httpListeners := make([]network.ApplicationGatewayHTTPListener, 0, len(configs))
+
+	for _, configRaw := range configs {
+		data := configRaw.(map[string]interface{})
+
+		name := data["name"].(string)
+		frontendIPConfigName := data["frontend_ip_configuration_name"].(string)
+		frontendIPConfigID := fmt.Sprintf("%s/frontendIPConfigurations/%s", gatewayID, frontendIPConfigName)
+		frontendPortName := data["frontend_port_name"].(string)
+		frontendPortID := fmt.Sprintf("%s/frontendPorts/%s", gatewayID, frontendPortName)
+		protocol := data["protocol"].(string)
+
+		listener := network.ApplicationGatewayHTTPListener{
+			Name: &name,
+			ApplicationGatewayHTTPListenerPropertiesFormat: &network.ApplicationGatewayHTTPListenerPropertiesFormat{
+				FrontendIPConfiguration: &network.SubResource{
+					ID: &frontendIPConfigID,
+				},
+				FrontendPort: &network.SubResource{
+					ID: &frontendPortID,
+				},
+				Protocol: network.ApplicationGatewayProtocol(protocol),
+			},
+		}
+
+		if host := data["host_name"].(string); host != "" {
+			listener.ApplicationGatewayHTTPListenerPropertiesFormat.HostName = &host
+		}
+
+		if sslCertName := data["ssl_certificate_name"].(string); sslCertName != "" {
+			certID := fmt.Sprintf("%s/sslCertificates/%s", gatewayID, sslCertName)
```

Comment:
given there's no guarantee that these URL's won't change in the future, would it make more sense to pass in the SSL Certificate ID, rather than the name?